### PR TITLE
Implement Image Update API

### PR DESCRIPTION
**Pull request by madushajg**

### CODE REVIEW

File: components/api/core/api_impl.bal [hunk 1]
```diff
@@ -499,30 +499,54 @@ returns http:Response {
...
-        sql:UpdateResult | error? updateImageRes = db:updateImage(orgName, imageName, updateImageBody.description, userId);
-
-        if (updateImageRes is sql:UpdateResult){
-            if (updateImageRes.updatedRowCount == 1) {
-                log:printDebug(io:sprintf("Image \'%s\' in organization \'%s\' is successfully updated. Author : %s", imageName, orgName, userId));
-                return buildSuccessResponse();
-            } else if (updateImageRes.updatedRowCount == 0){
-                log:printError(io:sprintf("Failed to update image \'%s\' in organization \'%s\' for Author %s : No matching records found",
-                imageName, orgName, userId));
+        log:printDebug(io:sprintf("Description : %s, Summary : %s", updateImageBody.description, updateImageBody.summary));
```
Comment (nadundesilva) [inline components/api/core/api_impl.bal]: Update this log to something more meaningful

File: components/api/core/api_impl.bal [hunk 2]
```diff
@@ -499,30 +499,54 @@ returns http:Response {
...
+            log:printDebug("Started transaction " + transactionId + " for updating image " + imageName);
+
+            sql:UpdateResult | error? updateImageRes = db:updateImageDescriptionNSummary(orgName, imageName, updateImageBody.description, updateImageBody.summary, userId);
+            if (updateImageRes is sql:UpdateResult) {
+                if (updateImageRes.updatedRowCount == 1) {
+                    log:printDebug(io:sprintf("Description and summery of the image %s/%s is successfully updated. Author : %s", orgName, imageName, userId));
+                    error? updateImageKeywordsRes = db:updateImageKeywords(orgName, imageName, updateImageBody.keywords, userId);
+                    if updateImageKeywordsRes is error {
+                        log:printError(io:sprintf("Failed to update image %s/%s for Author %s.", orgName, imageName, userId), err = updateImageKeywordsRes);
+                        resp = buildErrorResponse(http:EXPECTATION_FAILED_417, constants:API_ERROR_CODE, "Unable to update image", "");
```
Comment (nadundesilva) [inline components/api/core/api_impl.bal]: You should call "abort;" here

File: components/api/core/api_impl.bal [hunk 3]
```diff
@@ -499,30 +499,54 @@ returns http:Response {
...
+                    log:printError(io:sprintf("Failed to update image %s/%s for Author %s : More than one matching records found",
+                    imageName, orgName, userId));
+                    resp = buildErrorResponse(http:EXPECTATION_FAILED_417, constants:API_ERROR_CODE, "Unable to update image", "");
+                }
             } else {
-                log:printError(io:sprintf("Failed to update image \'%s\' in organization \'%s\' for Author %s : More than one matching records found",
-                imageName, orgName, userId));
+                log:printError(io:sprintf("Unexpected error occured while updating image %s/%s", imageName, orgName),
+                err = updateImageRes);
+                resp = buildUnknownErrorResponse();
```
Comment (nadundesilva) [inline components/api/core/api_impl.bal]: Better to call "abort;" here as well. (Although it won't make a big difference)

File: components/api/core/api_impl.bal [hunk 4]
```diff
@@ -499,30 +499,54 @@ returns http:Response {
...
+                        resp = buildSuccessResponse();
+                    }
+                } else if (updateImageRes.updatedRowCount == 0) {
+                    log:printError(io:sprintf("Failed to update image %s/%s for Author %s : No matching records found",
+                    imageName, orgName, userId));
+                    resp = buildErrorResponse(http:EXPECTATION_FAILED_417, constants:API_ERROR_CODE, "Unable to update image", "");
+                } else {
+                    log:printError(io:sprintf("Failed to update image %s/%s for Author %s : More than one matching records found",
+                    imageName, orgName, userId));
+                    resp = buildErrorResponse(http:EXPECTATION_FAILED_417, constants:API_ERROR_CODE, "Unable to update image", "");
```
Comment (nadundesilva) [inline components/api/core/api_impl.bal]: You should call "abort;" here

File: components/api/core/api_impl.bal [hunk 5]
```diff
@@ -499,30 +499,54 @@ returns http:Response {
...
+                        log:printError(io:sprintf("Failed to update image %s/%s for Author %s.", orgName, imageName, userId), err = updateImageKeywordsRes);
+                        resp = buildErrorResponse(http:EXPECTATION_FAILED_417, constants:API_ERROR_CODE, "Unable to update image", "");
+                    } else {
+                        log:printDebug(io:sprintf("Successfully updated the keywords of the image %s/%s by %s", orgName, imageName, userId));
+                        resp = buildSuccessResponse();
+                    }
+                } else if (updateImageRes.updatedRowCount == 0) {
+                    log:printError(io:sprintf("Failed to update image %s/%s for Author %s : No matching records found",
+                    imageName, orgName, userId));
+                    resp = buildErrorResponse(http:EXPECTATION_FAILED_417, constants:API_ERROR_CODE, "Unable to update image", "");
```
Comment (nadundesilva) [inline components/api/core/api_impl.bal]: You should call "abort;" here

File: components/api/core/api_impl.bal (hunk 2 above)
Comment (nadundesilva) [reply components/api/core/api_impl.bal]: IMO the error code should be 500. In the below error conditions as well.

File: components/api/core/api_impl.bal (hunk 5 above)
Comment (nadundesilva) [reply components/api/core/api_impl.bal]: IMO the error code should be 404.

File: components/api/core/api_impl.bal (hunk 4 above)
Comment (nadundesilva) [reply components/api/core/api_impl.bal]: IMO the error code should be 500.

File: components/api/db/cellery_hub_db.bal [hunk 6]
```diff
@@ -20,12 +20,23 @@ import ballerina/log;
...
+type RegistryArtifactImage record {|
+    string ARTIFACT_IMAGE_ID;
+|};
+
 public function getOrganization(string orgName) returns json | error {
     log:printDebug(io:sprintf("Performing data retreival on REGISTRY_ORGANIZATION table, Org name : \'%s\': ", orgName));
-    table<record {}> res =  check connection->select(GET_ORG_QUERY, gen:OrgResponse, orgName, loadToMemory = true);
+    table< record {
+
+}> res = check connection->select(GET_ORG_QUERY, gen:OrgResponse, orgName, loadToMemory = true);
```
Comment (nadundesilva) [inline components/api/db/cellery_hub_db.bal]: Shall we fix these formatting issues ? These seems too much to ignore till Ballerina tooling formatting fixes come.

File: components/api/db/cellery_hub_db.bal [hunk 7]
```diff
@@ -122,14 +139,20 @@ public function getImageKeywords(string imageId) returns table<gen:StringRecord>
...
     return buildJsonPayloadForGetArtifact(res, orgName, imageName, artifactVersion);
 }
 
-function buildJsonPayloadForGetArtifact(table<record {}> res, string orgName, string imageName, string artifactVersion) returns json | error {
+function buildJsonPayloadForGetArtifact(table< record {
+
+}> res, string orgName, string imageName, string artifactVersion) returns json | error {
     if (res.count() == 1) {
-        json resPayload = {};
+        json resPayload = {
```
Comment (nadundesilva) [inline components/api/db/cellery_hub_db.bal]: Let's revert this change.

File: components/api/db/cellery_hub_db.bal [hunk 8]
```diff
@@ -171,23 +194,30 @@ returns table<gen:Count> | error {
...
-    gen:OrgListResponse orgListResponse = {count:totalOrgs , data:[]};
-    if (totalOrgs > 0){
+    gen:OrgListResponse orgListResponse = {
+        count: totalOrgs,
+        data: []
+    };
+    if (totalOrgs > 0) {
         log:printDebug(io:sprintf("%d organization(s) found with the name \'%s\'", totalOrgs, orgName));
-        map<any> imageCountMap = {};
+        map<any> imageCountMap = {
```
Comment (nadundesilva) [inline components/api/db/cellery_hub_db.bal]: Let's revert this change.

File: components/api/db/cellery_hub_db.bal (hunk 7 above)
Comment (nadundesilva) [inline components/api/db/cellery_hub_db.bal]: Let's revert this change. Ideally an empty object should be on one line.

File: components/api/db/cellery_hub_db.bal [hunk 9]
```diff
@@ -203,33 +233,40 @@ public function searchUserOrganizations(string userId, string apiUserId, string
...
-    gen:OrgListResponse orgListResponse = {count:totalOrgs , data:[]};
-    if (totalOrgs > 0){
+    gen:OrgListResponse orgListResponse = {
+        count: totalOrgs,
+        data: []
+    };
+    if (totalOrgs > 0) {
         log:printDebug(io:sprintf("%d organization(s) found with the orgName \'%s\' for userId %s", totalOrgs, orgName, userId));
-        map<any> imageCountMap = {};
+        map<any> imageCountMap = {
```
Comment (nadundesilva) [inline components/api/db/cellery_hub_db.bal]: Let's revert this change. Ideally an empty object should be on one line.

File: components/api/core/api_impl.bal [hunk 10]
```diff
@@ -499,30 +499,54 @@ returns http:Response {
 # + orgName - Organization name
 # + imageName - Image ID
 # + return - http response (200 if success, 405 or 500 otherwise)
-public function updateImage (http:Request updateImageReq, string orgName, string imageName, gen:ImageUpdateRequest updateImageBody) returns http:Response {
+public function updateImage(http:Request updateImageReq, string orgName, string imageName, gen:ImageUpdateRequest updateImageBody) returns http:Response {
     if (updateImageReq.hasHeader(constants:AUTHENTICATED_USER)) {
```
Comment (nadundesilva) [inline components/api/core/api_impl.bal]: Need to check if the user has permission (in the org the image belongs to) to update the image.

File: components/api/db/cellery_hub_db.bal [hunk 11]
```diff
@@ -340,32 +389,95 @@ returns json | error {
...
+public function updateImageDescriptionNSummary(string orgName, string imageName, string description, string summary, string userId) returns sql:UpdateResult | error? {
+    log:printInfo(io:sprintf("Updating description and summary of the image %s/%s", orgName, imageName));
+    sql:UpdateResult res = check connection->update(UPDATE_IMAGE_DESCRIPTION_N_SUMMARY_QUERY, description, summary, imageName, orgName, userId);
     return res;
 }
 
+public function updateImageKeywords(string orgName, string imageName, string[] keywords, string userId) returns error? {
+    log:printInfo(io:sprintf("Updating keywords of the image %s/%s", orgName, imageName));
+    string role = check getUserRole(userId, orgName);
+    if role == constants:ROLE_ADMIN || role == constants:ROLE_PUSH {
```
Comment (nadundesilva) [inline components/api/db/cellery_hub_db.bal]: This validation should not be done at this level. This should be done at the API level

File: components/api/db/cellery_hub_db.bal [hunk 12]
```diff
@@ -340,32 +389,95 @@ returns json | error {
...
+            }
+        } else {
+            string errMsg = io:sprintf("Unable to update image %s/%s. Artifact Image Id of the image is not found", orgName, imageName);
+            log:printDebug(errMsg);
+            error er = error(errMsg);
+            return er;
+        }
+    } else {
+        string errMsg = io:sprintf("User %s is not allowd to update keywords of the image %s/%s", userId, orgName, imageName);
+        log:printDebug(errMsg);
```
Comment (nadundesilva) [inline components/api/db/cellery_hub_db.bal]: This should be an error log.

File: components/api/db/cellery_hub_db.bal [hunk 13]
```diff
@@ -340,32 +389,95 @@ returns json | error {
...
+            }
+            if i > 0 {
+                _ = check connection->batchUpdate(INSERT_IMAGE_KEYWORDS_QUERY, ...dataBatch);
+                log:printInfo(io:sprintf("Successfully inserted keywords for the image %s/%s", orgName, imageName));
+            } else {
+                log:printDebug(io:sprintf("No keywords found. Hence not performing keyword insertion for image %s/%s", orgName, imageName));
+            }
+        } else {
+            string errMsg = io:sprintf("Unable to update image %s/%s. Artifact Image Id of the image is not found", orgName, imageName);
+            log:printDebug(errMsg);
```
Comment (nadundesilva) [inline components/api/db/cellery_hub_db.bal]: This should be an error log

File: components/api/core/api_impl.bal (hunk 4 above)
Comment (madushajg) [reply components/api/core/api_impl.bal]: Since the error occurred because of not find any matching records, created a new API error code (ENTRY_NOT_FOUND_ERROR_CODE). Hence using buildErrorResponse()

File: components/api/core/api_impl.bal (hunk 5 above)
Comment (madushajg) [reply components/api/core/api_impl.bal]: Since the error occurred because of not find any matching records, created a new API error code (ENTRY_NOT_FOUND_ERROR_CODE). Hence using buildErrorResponse()

File: components/api/core/api_impl.bal (hunk 10 above)
Comment (madushajg) [reply components/api/core/api_impl.bal]: It checked within the query "UPDATE_IMAGE_DESCRIPTION_N_SUMMARY_QUERY"

File: components/api/db/cellery_hub_db.bal (hunk 11 above)
Comment (madushajg) [reply components/api/db/cellery_hub_db.bal]: This validation is done in the above-mentioned query. Hence removing this validation.

File: components/api/db/cellery_hub_db.bal (hunk 12 above)
Comment (madushajg) [reply components/api/db/cellery_hub_db.bal]: Remove this log because of skipping the role validation
